packages/push/examples/smallest/main.rs

```diff
@@ -0,0 +1,215 @@
+
+        if best.test_results.total_result.error == 0 {
+            println!("SUCCESS");
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+fn score_genome(
+    genome: &Plushy,
+    training_cases: &Cases<Input, Output>,
+) -> TestResults<test_results::Error<i128>> {
+    // The penalty value to use when an evolved program doesn't have an expected
+    // "return" value on the appropriate stack at the end of its execution.
+    const PENALTY_VALUE: i128 = 1_000;
+    let program = Vec::<PushProgram>::from(genome.clone());
+    let errors: TestResults<test_results::Error<i128>> = training_cases
+        .iter()
+        .map(
+            |&Case {
+                 input,
+                 output: Output(expected),
+             }: &Case<Input, Output>| {
+                run_case(&program, input, PENALTY_VALUE, expected)
+            },
+        )
+        .collect();
+    errors
+}
+
+fn run_case(program: &[PushProgram], input: Input, penalty_value: i128, expected: i64) -> i128 {
+    build_state(program, input).map_or(penalty_value, |start_state| {
+        // I don't think we're properly handling things like exceeding maximum
+        // stack size. I think the "Push way" here would be to take whatever
+        // value is on top of the relevant stack and go with it, but we instead
+        // return the penalty value.
+        start_state
+            .run_to_completion()
+            .map_or(penalty_value, |final_state| {
+                compute_error(&final_state, penalty_value, expected)
+            })
+    })
+}
+
+fn compute_error(final_state: &PushState, penalty_value: i128, expected: i64) -> i128 {
+    final_state
+        .stack::<i64>()
+        .top()
+        .map_or(penalty_value, |answer| {
+            i128::from(*answer)
+                .saturating_sub(i128::from(expected))
+                .abs()
+        })
+}
+
+fn build_state(program: &[PushProgram], Input(a, b, c, d): Input) -> Result<PushState, StackError> {
+    Ok(PushState::builder()
+        .with_max_stack_size(1000)
+        .with_program(program.to_vec())?
+        .with_int_input("a", a)
+        .with_int_input("b", b)
+        .with_int_input("c", c)
+        .with_int_input("d", d)
+        .build())
+}
+
+fn training_inputs(rng: &mut ThreadRng) -> Vec<Input> {
+    const NUM_TRAINING_CASES: usize = 100;
+    (0..NUM_TRAINING_CASES)
+        .map(|_| {
+            Input(
+                (rng.next_u32() % 100).into(),
+                (rng.next_u32() % 100).into(),
+                (rng.next_u32() % 100).into(),
+                (rng.next_u32() % 100).into(),
+            )
+        })
+        .collect::<Vec<_>>()
+}
+
+fn smallest(&Input(a, b, c, d): &Input) -> Output {
+    #[allow(clippy::unwrap_used)]
+    Output([a, b, c, d].into_iter().min().unwrap())
+}
+
+fn training_cases(rng: &mut ThreadRng) -> Cases<Input, Output> {
+    training_inputs(rng).with_target_fn(smallest)
+}
```

Comment:
and this I would probably inline as well, since this is so simple - no real readability advantage here, just needing to jump through a lot of functions